### PR TITLE
EOS-20682: Enhance sspl self-test to run test against physical server & storage enclosure

### PR DESCRIPTION
**Pull request by mariyappanp**

### CODE REVIEW

File: sspl_test/alerts/node/test_node_disk_actuator.py [hunk 1]
```diff
@@ -84,33 +82,5 @@ def test_node_disk_module_actuator(agrs):
             if "States Asserted" in disk_specific_infos:
                 assert(disk_specific_infos.get("States Asserted") is not None)
 
-def send_msg_request(resource_type, instance_id):
-    request = {
-	"username": "sspl-ll",
-	"expires": 3600,
-	"description": "Seagate Storage Platform Library - Actuator Request",
```
Comment (mariyappanp) [inline sspl_test/alerts/node/test_node_disk_actuator.py:91]: This has been moved to sspl_test/common.py for code resuability.

**Review by thavanathan — APPROVED** [review]
LGTM.

File: sspl_test/alerts/self_hw/self_hw_utilities.py [hunk 2]
```diff
@@ -63,3 +64,32 @@ def get_manufacturer_name():
...
+        "Product Name": None,
+        "Product Part Number": None
+        }
+    cmd = "ipmitool fru print"
+    search_res = ""
+    res_op, _, res_rc = SimpleProcess(cmd).run()
+    if isinstance(res_op, bytes):
+        res_op = res_op.decode("utf-8")
+    if res_rc == 0:
+        search_res = re.search(r"((.*[\S\n\s]+ID 1\)).*)|(.*[\S\n\s]+)", res_op)
```
Comment (cdeshmukh) [inline sspl_test/alerts/self_hw/self_hw_utilities.py:86]: We need to extract from "FRU Device (ID 0)", here seems you are look for ID1

Please add code comment here that code looking for 
FRU Device Description : Builtin FRU Device (ID 0)

File: sspl_test/alerts/self_hw/test_node_bmc.py [hunk 3]
```diff
@@ -19,38 +19,46 @@
...
-    pkey_file = "/root/.ssh/id_rsa"
-    if not os.path.exists(pkey_file):
-        print("\tSkipping test_bmc_config.. pkey file not found.")
-        return
+    """Check if BMC configuration are valid.
+
+    Testing BMC config with ipmitool is possible only when ipmi over lan
+    is configured(out-band setup). It is taken care by test_bmc_is_accessible.
+    So, validation on bmc onfiguration with bmc ip, user and secret value
+    through ssh is fine at this time.
```
Comment (cdeshmukh) [inline sspl_test/alerts/self_hw/test_node_bmc.py:51]: sure. no ssh key file requirement now !?

File: sspl_test/alerts/self_hw/test_node_bmc.py [hunk 4]
```diff
@@ -110,7 +118,46 @@ def test_bmc_is_accessible(args):
...
+    res_op, res_err, res_rc = SimpleProcess(ipmi_ver_cmd).run()
+    if res_rc == 0:
+        res_op = res_op.decode()
+        search_res = re.search(
+            r"%s[\s]+([\w.]+)(.*)" % IPMI_VERSION, res_op)
+        if search_res:
+            version_found = search_res.groups()[0]
+    else:
+        raise Exception("ERROR: %s" % res_err.decode())
+    if not (version_found == REQUIRED_IPMI_VERSION):
```
Comment (cdeshmukh) [inline sspl_test/alerts/self_hw/test_node_bmc.py:137]: sspl is fine with IPMI version >= 2.0

File: sspl_test/alerts/self_hw/self_hw_utilities.py (hunk 2 above)
Comment (mariyappanp) [reply sspl_test/alerts/self_hw/self_hw_utilities.py:86]: Regex here is to grep information between starting string to "ID 1".
So ideally ID 0 will be grepped.
Sure I will add the comment.

File: low-level/files/opt/seagate/sspl/setup/sspl_prepare.py [hunk 5]
```diff
@@ -111,6 +113,43 @@ def validate(self):
...
+                if not any(err for err in kcs_errors if err in res_err):
+                    raise Exception("BMC accessible through KCS - ERROR: %s" % res_err)
+        elif channel_interface == "lan":
+            # Check BMC is accessible through LAN
+            subcommand = "channel info"
+            cmd = "sudo ipmitool -H %s -U %s -P %s -I lan %s" %(
+                bmc_ip, bmc_user, bmc_passwd, subcommand)
+            _, res_err, res_rc = SimpleProcess(cmd).run()
+            if res_rc != 0:
+                raise Exception("BMC accessible over lan - ERROR: %s" % res_err.decode())
```
[the remark this answers is not in the document]
Comment (mariyappanp) [reply low-level/files/opt/seagate/sspl/setup/sspl_prepare.py:151]: ok, I will address this.

File: low-level/files/opt/seagate/sspl/setup/sspl_prepare.py [hunk 6]
```diff
@@ -111,6 +113,43 @@ def validate(self):
...
+                    channel_found = search_res.groups()[0]
+                if expected_channel != channel_found:
+                    print("UNEXPECTED BMC CHANNEL TYPE FOUND.")
+                    print("Expected: %s" % expected_channel)
+                    print("Found: %s" % channel_found)
+            else:
+                res_err = res_err.decode()
+                kcs_errors = ("could not find inband device", "driver timeout")
+                if not any(err for err in kcs_errors if err in res_err):
+                    raise Exception("BMC accessible through KCS - ERROR: %s" % res_err)
```
[the remark this answers is not in the document]
Comment (mariyappanp) [reply low-level/files/opt/seagate/sspl/setup/sspl_prepare.py:143]: Ok, I will address this.

File: sspl_test/alerts/self_hw/test_node_bmc.py (hunk 3 above)
Comment (mariyappanp) [reply sspl_test/alerts/self_hw/test_node_bmc.py:51]: Yes, I avoided that.
This test_bmc_config() is now just validating given BMC IP, user, passwd.
So ssh key file dependency is not required.

test_bmc_is_accessible() method is validating rc of ipmitool with that given config.
```cmd = "sudo ipmitool -H %s -U %s -P %s -I lan %s" %(
            bmc_ip, bmc_user, bmc_passwd, subcommand)
```

File: sspl_test/alerts/self_hw/test_node_bmc.py (hunk 4 above)
Comment (mariyappanp) [reply sspl_test/alerts/self_hw/test_node_bmc.py:137]: Ok, I will address this.

File: low-level/files/opt/seagate/sspl/setup/sspl_prepare.py (hunk 6 above)
[the remark this answers is not in the document]
Comment (mariyappanp) [reply low-level/files/opt/seagate/sspl/setup/sspl_prepare.py]: Done.

File: low-level/files/opt/seagate/sspl/setup/sspl_prepare.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (mariyappanp) [reply low-level/files/opt/seagate/sspl/setup/sspl_prepare.py]: Done.

File: sspl_test/alerts/self_hw/self_hw_utilities.py (hunk 2 above)
Comment (mariyappanp) [reply sspl_test/alerts/self_hw/self_hw_utilities.py]: Done.

File: sspl_test/alerts/self_hw/test_node_bmc.py (hunk 3 above)
Comment (mariyappanp) [reply sspl_test/alerts/self_hw/test_node_bmc.py:52]: Done.

File: sspl_test/alerts/self_hw/test_node_bmc.py (hunk 4 above)
Comment (mariyappanp) [reply sspl_test/alerts/self_hw/test_node_bmc.py]: Done.

File: sspl_test/alerts/self_hw/test_node_bmc.py (hunk 3 above)
Comment (cdeshmukh) [reply sspl_test/alerts/self_hw/test_node_bmc.py:52]: good

File: sspl_test/alerts/self_hw/self_hw_utilities.py (hunk 2 above)
Comment (cdeshmukh) [reply sspl_test/alerts/self_hw/self_hw_utilities.py]: Ok

**Review by cdeshmukh — APPROVED** [review]
After tmain is merged in main branch, rebase this patch with main branch later again

**Comment by cdeshmukh** [comment]
@mariyappanp codacy issues can you look into

**Review by gargpalak048 — COMMENTED** [review]
low-level/files/opt/seagate/sspl/setup/sspl_prepare.py
Use logger to log the statements instead of print statement.

**Review by sandeepsgt — APPROVED** [review]

**Review by gargpalak048 — APPROVED** [review]

**Comment by cdeshmukh** [comment]
VM deployment was done for this successfully before PR template update (#403) was merged. Merging this too

**Pull request merged**